contracts/chain-adapters/Solana_Adapter.sol

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { IMessageTransmitter, ITokenMessenger } from "../external/interfaces/CCTPInterfaces.sol";
+import { SpokePoolInterface } from "../interfaces/SpokePoolInterface.sol";
+import { AdapterInterface } from "./interfaces/AdapterInterface.sol";
+import { CircleCCTPAdapter, CircleDomainIds } from "../libraries/CircleCCTPAdapter.sol";
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @notice Contract containing logic to send messages from L1 to Solana.
+ * @dev Public functions calling external contracts do not guard against reentrancy because they are expected to be
+ * called via delegatecall, which will execute this contract's logic within the context of the originating contract.
+ * For example, the HubPool will delegatecall these functions, therefore it's only necessary that the HubPool's methods
+ * that call this contract's logic guard against reentrancy.
+ * @custom:security-contact bugs@across.to
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Solana_Adapter is AdapterInterface, CircleCCTPAdapter {
+    /**
+     * @notice The official Circle CCTP MessageTransmitter contract endpoint.
+     * @dev Posted officially here: https://developers.circle.com/stablecoins/docs/evm-smart-contracts
+     */
+    // solhint-disable-next-line immutable-vars-naming
+    IMessageTransmitter public immutable cctpMessageTransmitter;
+
+    // Solana spoke pool address, decoded from Base58 to bytes32.
+    bytes32 public immutable SOLANA_SPOKE_POOL_BYTES32;
+
+    // Solana spoke pool address, mapped to its EVM address representation.
+    address public immutable SOLANA_SPOKE_POOL_ADDRESS;
+
+    // USDC mint address on Solana, decoded from Base58 to bytes32.
+    bytes32 public immutable SOLANA_USDC_BYTES32;
+
+    // USDC mint address on Solana, mapped to its EVM address representation.
+    address public immutable SOLANA_USDC_ADDRESS;
+
+    // USDC token address on Solana for the spoke pool (vault ATA), decoded from Base58 to bytes32.
+    bytes32 public immutable SOLANA_SPOKE_POOL_USDC_VAULT;
+
+    // Custom errors for constructor argument validation.
+    error InvalidCctpTokenMessenger(address tokenMessenger);
+    error InvalidCctpMessageTransmitter(address messageTransmitter);
+
+    // Custom errors for relayMessage validation.
+    error InvalidRelayMessageTarget(address target);
+    error InvalidOriginToken(address originToken);
+    error InvalidDestinationChainId(uint256 destinationChainId);
+
+    // Custom errors for relayTokens validation.
+    error InvalidL1Token(address l1Token);
+    error InvalidL2Token(address l2Token);
+    error InvalidAmount(uint256 amount);
+    error InvalidTokenRecipient(address to);
```

Comment:
I used these for debugging, though HubPool does not bubble up revert data on delegatecall. Shall we still keep custom errors or revert with empty data at the adapter?

Reply:
I think they are ok to have adapter specific errors? if there are other errors that other adapters re-use we should cleary use them but things like `InvalidCctpTokenMessenger` only apply in this context.